File: ts/src/ollama/runner.ts
import { createInterface } from "node:readline";
import { Readable } from "node:stream";
import type { Runner, Result, Message, Session, Usage } from "../types.js";
import {
  NotFoundError,
  TimeoutError,
  CancelledError,
  ParseError,
  NoResultError,
  NonZeroExitError,
} from "../errors.js";
import type {
  OllamaRunnerConfig,
  OllamaRunOptions,
  FetchFn,
} from "./options.js";
import type {
  ChatRequest,
  ChatMessage,
  ChatResponse,
  ModelOptions,
} from "./types.js";

const DEFAULT_BASE_URL = "http://localhost:11434";

/** Create an Ollama runner. */
export function createOllamaRunner(config: OllamaRunnerConfig = {}): Runner {
  const baseURL = config.baseURL ?? DEFAULT_BASE_URL;
  const fetchFn: FetchFn = config.fetch ?? fetch;

  return {
    start: (prompt, options) =>
      start(config, fetchFn, baseURL, prompt, options as OllamaRunOptions),
    run: (prompt, options) =>
      run(config, fetchFn, baseURL, prompt, options as OllamaRunOptions),
    runStream: (prompt, options) =>
      runStream(config, fetchFn, baseURL, prompt, options as OllamaRunOptions),
  };
}

function start(
  config: OllamaRunnerConfig,
  fetchFn: FetchFn,
  baseURL: string,
  prompt: string,
  options: OllamaRunOptions = {},
): Session {
  const { signal, clearTimeout: clearTO } = combinedSignal(options);
  const abortController = new AbortController();

  // Chain: if the combined signal fires, abort the controller too.
  signal.addEventListener("abort", () => abortController.abort(signal.reason), {
    once: true,
  });

  let resolveResult: (value: Result) => void;
  let rejectResult: (reason: unknown) => void;
  const resultPromise = new Promise<Result>((resolve, reject) => {
    resolveResult = resolve;
    rejectResult = reject;
  });

  async function* messageGenerator(): AsyncGenerator<Message> {
    try {
      const body = buildRequestBody(prompt, options);

      logRequest(config, baseURL);

      let resp: Response;
      try {
        resp = await fetchFn(`${baseURL}/api/chat`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body),
          signal: abortController.signal,
        });
      } catch (err) {
        // Check if this was caused by our signal (timeout or cancel).
        if (signal.aborted) {
          throw abortError(signal);
        }
        throw mapFetchError(err);
      }

      if (resp.status === 404) {
        throw new NotFoundError("model not found (HTTP 404)");
      }
      if (!resp.ok) {
        throw new NonZeroExitError(resp.status, `HTTP ${resp.status}`);
      }

      if (!resp.body) {
        throw new NoResultError();
      }

      const nodeStream = Readable.fromWeb(
        resp.body as import("node:stream/web").ReadableStream,
      );
      const rl = createInterface({ input: nodeStream });

      const textParts: string[] = [];
      let finalResp: ChatResponse | undefined;

      for await (const line of rl) {
        if (abortController.signal.aborted) break;
        if (!line) continue;

        let chunk: ChatResponse;
        try {
          chunk = JSON.parse(line) as ChatResponse;
        } catch {
          throw new ParseError(`invalid JSON: ${line}`);
        }

        if (chunk.message.content) {
          textParts.push(chunk.message.content);
        }

        if (chunk.done) {
          finalResp = chunk;
        }

        const msg: Message = {
          type: chunk.done ? "result" : "assistant",
          raw: line,
        };

        if (options.onMessage) {
          options.onMessage(msg);
        }

        yield msg;
      }

      clearTO();

      if (abortController.signal.aborted) {
        const err = abortError(signal);
        rejectResult!(err);
        return;
      }

      if (!finalResp) {
        rejectResult!(new NoResultError());
        return;
      }

      const usage: Usage = {
        inputTokens: finalResp.prompt_eval_count ?? 0,
        outputTokens: finalResp.eval_count ?? 0,
        cacheCreationInputTokens: 0,
        cacheReadInputTokens: 0,
      };

      resolveResult!({
        text: textParts.join(""),
        isError: false,
        exitCode: 0,
        usage,
        costUSD: 0,
        durationMs: finalResp.total_duration
          ? Math.floor(finalResp.total_duration / 1e6)
          : 0,
        sessionId: "",
      });
    } catch (err) {
      clearTO();
      if (err instanceof NotFoundError || err instanceof NonZeroExitError || err instanceof ParseError || err instanceof NoResultError) {
        rejectResult!(err);
        return;
      }
      rejectResult!(mapFetchError(err));
    }
  }

  const messages = messageGenerator();

  return {
    messages,
    result: resultPromise,
    abort: () => {
      abortController.abort("cancelled");
    },
    send: () => {
      throw new Error("not yet supported");
    },
  };
}

async function run(
  config: OllamaRunnerConfig,
  fetchFn: FetchFn,
  baseURL: string,
  prompt: string,
  options: OllamaRunOptions = {},
): Promise<Result> {
  const session = start(config, fetchFn, baseURL, prompt, options);

  for await (const _msg of session.messages) {
    // consumed
  }

  return session.result;
}

async function* runStream(
  config: OllamaRunnerConfig,
  fetchFn: FetchFn,
  baseURL: string,
  prompt: string,
  options: OllamaRunOptions = {},
): AsyncGenerator<Message> {
  const session = start(config, fetchFn, baseURL, prompt, options);
  yield* session.messages;

  // Propagates timeout/cancel/parse errors to the stream consumer.
  await session.result;
}

function buildRequestBody(
  prompt: string,
  options: OllamaRunOptions,
): ChatRequest {
  const messages = buildMessages(prompt, options);

  const req: ChatRequest = {
    model: options.model ?? "",
    messages,
    stream: true,
  };

  if (options.think != null) {
    req.think = options.think;
  }
  if (options.format) {
    req.format = options.format;
  }
  if (options.keepAlive) {
    req.keep_alive = options.keepAlive;
  }

  const modelOpts = buildModelOptions(options);
  if (modelOpts) {
    req.options = modelOpts;
  }

  return req;
}

function buildMessages(
  prompt: string,
  options: OllamaRunOptions,
): ChatMessage[] {
  const messages: ChatMessage[] = [];

  let systemPrompt = options.systemPrompt ?? "";
  if (options.appendSystemPrompt) {
    if (systemPrompt) {
      systemPrompt += "\n" + options.appendSystemPrompt;
    } else {
      systemPrompt = options.appendSystemPrompt;
    }
  }

  if (systemPrompt) {
    messages.push({ role: "system", content: systemPrompt });
  }

  messages.push({ role: "user", content: prompt });
  return messages;
}

function buildModelOptions(options: OllamaRunOptions): ModelOptions | undefined {
  const hasAny =
    options.temperature != null ||
    options.numCtx != null ||
    options.numPredict != null ||
    options.seed != null ||
    options.stop != null ||
    options.topK != null ||
    options.topP != null ||
    options.minP != null;

  if (!hasAny) return undefined;

  const opts: ModelOptions = {};
  if (options.temperature != null) opts.temperature = options.temperature;
  if (options.numCtx != null) opts.num_ctx = options.numCtx;
  if (options.numPredict != null) opts.num_predict = options.numPredict;
  if (options.seed != null) opts.seed = options.seed;
  if (options.stop != null) opts.stop = options.stop;
  if (options.topK != null) opts.top_k = options.topK;
  if (options.topP != null) opts.top_p = options.topP;
  if (options.minP != null) opts.min_p = options.minP;
  return opts;
}

/** Combine timeout and user-provided signal into a single AbortSignal. */
function combinedSignal(
  options: OllamaRunOptions,
): { signal: AbortSignal; clearTimeout: () => void } {
  const signals: AbortSignal[] = [];
  let timeoutId: ReturnType<typeof setTimeout> | undefined;

  if (options.signal) {
    signals.push(options.signal);
  }

  if (options.timeout != null && options.timeout > 0) {
    const controller = new AbortController();
    timeoutId = setTimeout(() => controller.abort("timeout"), options.timeout);
    signals.push(controller.signal);
  }

  if (signals.length === 0) {
    const controller = new AbortController();
    return { signal: controller.signal, clearTimeout: () => {} };
  }

  if (signals.length === 1) {
    return {
      signal: signals[0],
      clearTimeout: () => {
        if (timeoutId != null) clearTimeout(timeoutId);
      },
    };
  }

  const combined = AbortSignal.any(signals);
  return {
    signal: combined,
    clearTimeout: () => {
      if (timeoutId != null) clearTimeout(timeoutId);
    },
  };
}

function abortError(signal: AbortSignal): TimeoutError | CancelledError {
  if (signal.reason === "timeout") {
    return new TimeoutError("execution timed out");
  }
  return new CancelledError("execution cancelled");
}

function mapFetchError(err: unknown): Error {
  if (err instanceof TimeoutError || err instanceof CancelledError || err instanceof NotFoundError || err instanceof NonZeroExitError || err instanceof ParseError || err instanceof NoResultError) {
    return err;
  }

  if (err instanceof DOMException || (err instanceof Error && err.name === "AbortError")) {
    // Cannot distinguish timeout from cancel via DOMException alone;
    // caller must check the signal reason before calling mapFetchError.
    return new CancelledError("execution cancelled");
  }

  if (err instanceof TypeError) {
    // fetch throws TypeError for network errors (connection refused, DNS failure, etc.)
    return new NotFoundError(`connection failed: ${err.message}`);
  }

  return new NotFoundError(`request failed: ${err}`);
}

function logRequest(config: OllamaRunnerConfig, baseURL: string): void {
  if (!config.logger) return;
  config.logger.debug("executing Ollama API request", {
    method: "POST",
    url: `${baseURL}/api/chat`,
  });
}

File: python/src/agentrunner/ollama/runner.py
"""Ollama runner implementation using the Ollama HTTP API."""

from __future__ import annotations

import asyncio
import json
from collections.abc import AsyncIterator
from http import HTTPStatus
from typing import Any

from ..errors import (
    CancelledError,
    NoResultError,
    NotFoundError,
    ParseError,
    RunnerError,
    TimeoutError,
)
from ..types import Message, Result, Usage
from .options import OllamaRunnerConfig, OllamaRunOptions
from .types import ChatResponse, ModelOptions

DEFAULT_BASE_URL = "http://localhost:11434"


class OllamaSession:
    """Session encapsulates a running Ollama API request.

    Supports ``async for msg in session`` to iterate messages,
    and ``await session.result`` to get the final result.
    """

    def __init__(
        self,
        config: OllamaRunnerConfig,
        prompt: str,
        options: OllamaRunOptions,
    ) -> None:
        self._config = config
        self._prompt = prompt
        self._options = options

        self._loop = asyncio.get_running_loop()
        self._queue: asyncio.Queue[Message | None] = asyncio.Queue()
        self._result_future: asyncio.Future[Result] = self._loop.create_future()
        self._aborted = False
        self._task: asyncio.Task[None] = asyncio.ensure_future(self._run_request())

    async def _run_request(self) -> None:
        try:
            base_url = self._config.base_url or DEFAULT_BASE_URL
            body = _build_request_body(self._prompt, self._options)

            if self._config.logger:
                self._config.logger.debug(
                    "executing Ollama API request",
                    extra={"method": "POST", "url": f"{base_url}/api/chat"},
                )

            reader, writer = await _http_post_stream(
                base_url, "/api/chat", json.dumps(body)
            )

            text_parts: list[str] = []
            final_resp: ChatResponse | None = None

            try:
                while True:
                    if self._aborted:
                        break

                    raw_line = await reader.readline()
                    if not raw_line:
                        break

                    line = raw_line.decode("utf-8", errors="replace").strip()
                    if not line:
                        continue

                    try:
                        chunk = ChatResponse.from_dict(json.loads(line))
                    except (json.JSONDecodeError, KeyError) as exc:
                        raise ParseError(f"invalid JSON: {line}") from exc

                    if chunk.message.content:
                        text_parts.append(chunk.message.content)

                    if chunk.done:
                        final_resp = chunk

                    msg = Message(
                        type="result" if chunk.done else "assistant",
                        raw=line,
                    )
                    await self._queue.put(msg)
            finally:
                writer.close()
                try:
                    await writer.wait_closed()
                except Exception:
                    pass

            if self._aborted:
                self._result_future.set_exception(CancelledError("execution cancelled"))
                return

            if not final_resp:
                self._result_future.set_exception(NoResultError())
                return

            usage = Usage(
                input_tokens=final_resp.prompt_eval_count or 0,
                output_tokens=final_resp.eval_count or 0,
            )

            self._result_future.set_result(
                Result(
                    text="".join(text_parts),
                    is_error=False,
                    exit_code=0,
                    usage=usage,
                    cost_usd=0.0,
                    duration_ms=(
                        final_resp.total_duration / 1e6
                        if final_resp.total_duration
                        else 0.0
                    ),
                    session_id="",
                )
            )
        except (CancelledError, TimeoutError, NotFoundError, ParseError, NoResultError):
            if not self._result_future.done():
                self._result_future.set_exception(
                    __import__("sys").exc_info()[1]  # type: ignore[arg-type]
                )
        except OSError as exc:
            err = NotFoundError(f"connection failed: {exc}")
            if not self._result_future.done():
                self._result_future.set_exception(err)
        except Exception as exc:
            if not self._result_future.done():
                self._result_future.set_exception(RunnerError(str(exc)))
        finally:
            await self._queue.put(None)

    def __aiter__(self) -> AsyncIterator[Message]:
        return self._message_iter()

    async def _message_iter(self) -> AsyncIterator[Message]:
        while True:
            msg = await self._queue.get()
            if msg is None:
                break
            yield msg

    @property
    def result(self) -> asyncio.Future[Result]:
        return self._result_future

    def abort(self) -> None:
        self._aborted = True

    def send(self, input: Any) -> None:
        raise NotImplementedError("send is not supported for Ollama runner")


class OllamaRunner:
    """Ollama runner — talks to the Ollama HTTP API.

    Construct directly::

        runner = OllamaRunner()
        runner = OllamaRunner(config=OllamaRunnerConfig(base_url="http://...", logger=logger))
    """

    def __init__(self, config: OllamaRunnerConfig | None = None) -> None:
        self._config = config or OllamaRunnerConfig()

    def start(
        self,
        prompt: str,
        options: OllamaRunOptions | None = None,
    ) -> OllamaSession:
        opts = options or OllamaRunOptions()
        if not opts.model:
            raise RunnerError("model is required for Ollama runner")

        timeout = opts.timeout
        session = OllamaSession(self._config, prompt, opts)

        if timeout is not None and timeout > 0:
            loop = asyncio.get_running_loop()
            loop.call_later(timeout, session.abort)

        return session

    async def run(
        self,
        prompt: str,
        options: OllamaRunOptions | None = None,
    ) -> Result:
        session = self.start(prompt, options)
        async for _msg in session:
            pass
        return await session.result

    async def run_stream(
        self,
        prompt: str,
        options: OllamaRunOptions | None = None,
    ) -> OllamaSession:
        return self.start(prompt, options)


async def _http_post_stream(
    base_url: str, path: str, body: str
) -> tuple[asyncio.StreamReader, asyncio.StreamWriter]:
    """Open a raw HTTP POST connection and return the response body stream.

    Uses asyncio streams directly to avoid external dependencies.
    Raises NotFoundError on connection failure or HTTP 404.
    """
    from urllib.parse import urlparse

    parsed = urlparse(base_url)
    host = parsed.hostname or "localhost"
    port = parsed.port or 80
    use_ssl = parsed.scheme == "https"

    try:
        if use_ssl:
            import ssl

            ctx = ssl.create_default_context()
            reader, writer = await asyncio.open_connection(host, port, ssl=ctx)
        else:
            reader, writer = await asyncio.open_connection(host, port)
    except OSError as exc:
        raise NotFoundError(f"connection failed: {exc}") from exc

    # Send HTTP request.
    body_bytes = body.encode("utf-8")
    request_lines = (
        f"POST {path} HTTP/1.1\r\n"
        f"Host: {host}:{port}\r\n"
        f"Content-Type: application/json\r\n"
        f"Content-Length: {len(body_bytes)}\r\n"
        f"Connection: close\r\n"
        f"\r\n"
    )
    writer.write(request_lines.encode("utf-8"))
    writer.write(body_bytes)
    await writer.drain()

    # Read status line.
    status_line = await reader.readline()
    status_str = status_line.decode("utf-8", errors="replace").strip()
    parts = status_str.split(" ", 2)
    if len(parts) < 2:
        raise NotFoundError(f"invalid HTTP response: {status_str}")

    status_code = int(parts[1])

    # Read headers (discard them, we just need the body stream).
    while True:
        header_line = await reader.readline()
        if header_line in (b"\r\n", b"\n", b""):
            break

    if status_code == HTTPStatus.NOT_FOUND:
        raise NotFoundError("model not found (HTTP 404)")
    if status_code >= 400:
        raise RunnerError(f"HTTP {status_code}")

    return reader, writer


def _build_request_body(prompt: str, options: OllamaRunOptions) -> dict[str, Any]:
    """Build the JSON request body for POST /api/chat."""
    messages: list[dict[str, str]] = []

    system_prompt = options.system_prompt or ""
    if options.append_system_prompt:
        if system_prompt:
            system_prompt += "\n" + options.append_system_prompt
        else:
            system_prompt = options.append_system_prompt

    if system_prompt:
        messages.append({"role": "system", "content": system_prompt})

    messages.append({"role": "user", "content": prompt})

    body: dict[str, Any] = {
        "model": options.model,
        "messages": messages,
        "stream": True,
    }

    if options.think is not None:
        body["think"] = options.think
    if options.format:
        body["format"] = options.format
    if options.keep_alive:
        body["keep_alive"] = options.keep_alive

    model_opts = _build_model_options(options)
    if model_opts:
        body["options"] = model_opts

    return body


def _build_model_options(options: OllamaRunOptions) -> dict[str, Any] | None:
    """Build the model options dict, or None if no options are set."""
    opts = ModelOptions(
        temperature=options.temperature,
        num_ctx=options.num_ctx,
        num_predict=options.num_predict,
        seed=options.seed,
        stop=options.stop,
        top_k=options.top_k,
        top_p=options.top_p,
        min_p=options.min_p,
    )
    d = opts.to_dict()
    return d if d else None
